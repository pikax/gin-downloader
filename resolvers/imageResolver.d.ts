import { IImageResolver, ImageCollection, IMangaConfigDependency, IMangaParserDependency, IMangaRequestFactoryDependency } from "../interface";
export declare type ImageResolverDependencies = IMangaParserDependency & IMangaConfigDependency & IMangaRequestFactoryDependency;
export declare class ImageResolver implements IImageResolver {
    private _parser;
    private _config;
    private _requestFactory;
    constructor(dependencies: ImageResolverDependencies);
    images(chapterSrc: any): Promise<ImageCollection>;
}
