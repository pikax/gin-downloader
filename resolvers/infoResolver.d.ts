import { IMangaConfigDependency, IMangaInfoResolver, IMangaParserDependency, IMangaRequestFactoryDependency, MangaInfo } from "../interface";
export declare type InfoResolverDependencies = IMangaParserDependency & IMangaConfigDependency & IMangaRequestFactoryDependency;
export declare class InfoResolver implements IMangaInfoResolver {
    private _parser;
    private _config;
    private _requestFactory;
    constructor(dependencies: InfoResolverDependencies);
    info(src: any): Promise<MangaInfo>;
}
