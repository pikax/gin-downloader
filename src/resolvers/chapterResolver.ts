import {ChapterSource, IChapterResolver, IMangaParserDependency, IMangaRequestFactoryDependency} from "../interface";
import {IMangaParser, IMangaRequestFactory} from "../manga/interface";



type di = IMangaParserDependency & IMangaRequestFactoryDependency;

export class ChapterResolver implements IChapterResolver {
    private _parser: IMangaParser;
    private _requestFactory: IMangaRequestFactory;

    constructor(dependencies: di) {
        this._parser = dependencies.parser;
        this._requestFactory = dependencies.requestFactory;
    }

    async chapters(src: string): Promise<ChapterSource[]> {
        const result = await this._requestFactory.request({uri: src});
        return Array.from(this._parser.chapters(result));
    }
}


