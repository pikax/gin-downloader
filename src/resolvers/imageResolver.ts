import {
    IImageResolver,
    ImageCollection, ImageSource,
    IMangaConfigDependency,
    IMangaParserDependency,
    IMangaRequestFactoryDependency
} from "../interface";
import {IMangaConfig, IMangaParser, IMangaRequestFactory} from "../manga/interface";
import {Lazy} from "../util/lazy";
import * as url from "url";


export type ImageResolverDependencies = IMangaParserDependency & IMangaConfigDependency & IMangaRequestFactoryDependency;


export class ImageResolver implements IImageResolver {
    private _parser: IMangaParser;
    private _config: IMangaConfig;
    private _requestFactory: IMangaRequestFactory;

    constructor(dependencies: ImageResolverDependencies) {
        this._parser = dependencies.parser;
        this._config = dependencies.config;
        this._requestFactory = dependencies.requestFactory;
    }

    async images(chapterSrc): Promise<ImageCollection> {
        let href = url.resolve(this._config.site, chapterSrc).toString();
        let result = await this._requestFactory.request({uri: href});

        const paths = this._parser.imagesPaths(result);
        paths.next(); // skip first, because we requested that already

        let src = this._parser.image(result);

        const promises = [
            new Lazy<Promise<ImageSource>>(() => Promise.resolve({
                name: url.parse(src).pathname.split("/").reverse()[0],
                src
            }))
        ];


        const promiseRequest = async (s): Promise<ImageSource> => {
            const h = url.resolve(href, s).toString();
            const result = await this._requestFactory.request({uri: h})
            const src = this._parser.image(result);

            return {
                name: url.parse(src).pathname.split("/").reverse()[0],
                src
            };
        };


        for (const it of paths) {
            promises.push(new Lazy<Promise<ImageSource>>(() => promiseRequest(it.src)));
        }


        return promises;
    }

}
