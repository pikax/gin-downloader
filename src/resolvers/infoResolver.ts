import {
    IMangaConfigDependency,
    IMangaInfoResolver,
    IMangaParserDependency,
    IMangaRequestFactoryDependency,
    Info,
    MangaInfo
} from "../interface";
import {IMangaConfig, IMangaParser, IMangaRequestFactory} from "../manga/interface";


type di = IMangaParserDependency & IMangaConfigDependency & IMangaRequestFactoryDependency;

export class InfoResolver implements IMangaInfoResolver {
    private _parser: IMangaParser;
    private _config: IMangaConfig;
    private _requestFactory: IMangaRequestFactory;

    constructor(dependencies: di) {
        this._parser = dependencies.parser;
        this._config = dependencies.config;
        this._requestFactory = dependencies.requestFactory;
    }

    async info(src): Promise<MangaInfo> {
        return this._requestFactory.request({uri: src})
            .then(this._parser.info);
    }

}