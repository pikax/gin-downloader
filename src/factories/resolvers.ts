import {IMangaDependencies, IMangaResolvers, IMangaResolversFactory} from "../interface";
import {ImageResolver} from "../resolvers/imageResolver";
import {ChapterResolver} from "../resolvers/chapterResolver";
import {InfoResolver} from "../resolvers/infoResolver";


export class DefaultResolverFactory implements IMangaResolversFactory {
    build(di: IMangaDependencies): IMangaResolvers {

        const imageResolver = new ImageResolver(di);
        const chapterResolver = new ChapterResolver(di);
        const infoResolver = new InfoResolver(di);

        return {
            imageResolver,
            chapterResolver,
            infoResolver,
        };
    }
}

