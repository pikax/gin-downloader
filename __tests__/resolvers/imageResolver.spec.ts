import {resolve} from 'url';
import {ImageResolver} from "../../src/resolvers/imageResolver";

describe("Info resolver", () => {
    it("Should resolve and parse info", async () => {

        const site = "http://test.test";

        const paths = [1, 2, 3, 4, 5];

        const imageResult = site + "/1/1";

        const gen = (function* () {
            yield* paths.map(x => ({src: resolve(imageResult, x.toString())}));
        })();

        const mockConfig = jest.fn();
        const mockRequest = jest.fn(({uri}: { uri: string }) => {
            return uri.toString().split("/").reverse()[0];
        });
        const mockImage = jest.fn((re) => resolve(imageResult, re));

        const mockImagePaths = jest.fn(() => {
            return gen;
        });


        const config = {name: mockConfig, latestUrl: mockConfig, mangasUrl: mockConfig, site: site} as any;
        const requestFactory = {request: mockRequest};
        const parser = {

            image: mockImage,

            imagesPaths: mockImagePaths,

        } as any;

        const info = new ImageResolver({config, requestFactory, parser});


        const src = "http://test.test/1";
        const i = await info.images(src);


        // expect(i).toBe(parseResult);

        expect(mockRequest.mock.calls.length).toBe(1);

        expect(mockRequest.mock.calls[0][0]).toBeDefined();
        expect(mockRequest.mock.calls[0][0].uri).toBe(src);

        expect(mockImage.mock.calls.length).toBe(1);
        expect(mockImage.mock.calls[0][0]).toBe("1");

        expect(mockConfig.mock.calls.length).toBe(0);


        const results = await Promise.all(i.map(x => x.value));


        expect(results).toMatchSnapshot();

    });

});
