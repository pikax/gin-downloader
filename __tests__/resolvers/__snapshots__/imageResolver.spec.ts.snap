// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info resolver Should resolve and parse info 1`] = `
Array [
  Object {
    "name": "1",
    "src": "http://test.test/1/1",
  },
  Object {
    "name": "2",
    "src": "http://test.test/1/2",
  },
  Object {
    "name": "3",
    "src": "http://test.test/1/3",
  },
  Object {
    "name": "4",
    "src": "http://test.test/1/4",
  },
  Object {
    "name": "5",
    "src": "http://test.test/1/5",
  },
]
`;
